go/vt/tabletserver/planbuilder/dml.go

```diff
@@ -323,12 +328,29 @@ func analyzeInsert(ins *sqlparser.Insert, getTable TableGetter) (plan *ExecPlan,
 	upd := &sqlparser.Update{
 		Comments: ins.Comments,
 		Table:    ins.Table,
-		Exprs:    sqlparser.UpdateExprs(ins.OnDup),
+		Exprs:    updateExprs,
 	}
 	plan.UpsertQuery = GenerateUpdateOuterQuery(upd)
 	return plan, nil
 }
 
+// resolveUpsertUpdateValues walks the UpdateExprs tree for an upsert, replacing
+// any VALUES(foo) expressions with the correct value from the rowList, in this
+// example the value for column 'foo'
+func resolveUpsertUpdateValues(rowList sqlparser.ValTuple, columns sqlparser.Columns, dup sqlparser.OnDup) (sqlparser.UpdateExprs, error) {
```

Comment:
Much simpler implementation, using Walk, which supports all of our complicated constructs.